README.md

```diff
@@ -1,132 +1,355 @@
+
+    The project uses the Streaming Inserts API of BigQuery which by default enables best-effort deduplication mechanism. However, you should not rely on this mechanism to guarantee the absence of duplicates in your data. For a possible solution, see the knowledge base article, checkout [high number of duplicates in Dataflow pipeline streaming inserts to BigQuery](https://cloud.google.com/knowledge/kb/high-number-of-duplicates-in-dataflow-pipeline-streaming-inserts-to-bigquery-000004276?authuser=0).
+
+
+2. Error in DLP API : "Too many findings in request"
+
+    DLP has a max findings per request [limit](https://cloud.google.com/dlp/limits#content-redaction-limits) of 3000. Run the pipeline again with a smaller batch size.
+
+
+3. DLP API request quota exhausted
+
+    This can happen if the Dataflow pipeline is being run with a small batch size. Rerun the pipeline with a larger batch size.
+
+    If the batch size cannot be increased or if the issue persists after you saturate the batch size:
+
+   * Increase the value of the dlpApiRetryCount parameter
+
+   * The dlp-dataflow-deidentification repo offers a parameter numShardsPerDLPRequestBatching. Reducing this below the default (100) will reduce the number of parallel requests sent to DLP.
+
+   * Check whether there are other pipelines/clients generating DLP API requests.
+
+   * Submit a request to increase the quota.
+
+
 ## Dataflow DAG
 
-For Deid and Inspect:
+For inspection and de-identification:
+![v2_dag_](diagrams/dlp_dag_new.png)
 
-![v2_dag_](diagrams/dlp_dag_new.png)	   	
 
 
+For re-identification
 
-For Reid:
+![v2_dag_](diagrams/dlp_reid_dag.png)
 
-![v2_dag_](diagrams/dlp_reid_dag.png)	   	
 
 
+## Advanced topics
 
-## Trigger Pipeline Using Public Image
-You can use the gcloud command to trigger the pipeline using Dataflow flex template. Below is an example for de-identification transform from a S3 bucket.
+You can use the gcloud command to trigger the pipeline using a Dataflow flex template. The following example runs a de-identification transformation on data from an S3 bucket.
 
 ```
 gcloud beta dataflow flex-template run "dlp-s3-scanner-deid-demo" --project=<project_id> \
 --region=<region> --template-file-gcs-location=gs://dataflow-dlp-solution-sample-data/dynamic_template_dlp_v2.json \
 --parameters=^~^streaming=true~enableStreamingEngine=true~tempLocation=gs://<path>/temp~numWorkers=5~maxNumWorkers=5~runner=DataflowRunner~filePattern=<s3orgcspath>/filename.csv~dataset=<bq_dataset>~autoscalingAlgorithm=THROUGHPUT_BASED~workerMachineType=n1-highmem-8~inspectTemplateName=<inspect_template>~deidentifyTemplateName=<deid_template>~awsRegion=ca-central-1~awsCredentialsProvider=$AWS_CRED~batchSize=100000~DLPMethod=DEID
 
 ```
-## To Do
-- take out first row as header before processing 
 
+To avoid repeatedly running Gradle, you can [create a flex template](https://cloud.google.com/dataflow/docs/guides/templates/using-flex-templates).
+
```

Comment:
I think this section should be:
1. You can create a flex template and use the template instead of having to build the template every time. Instructions to create flex template<link>.
2. gcloud CLI can be used to run a pipeline from a flex template <link>.

Reply:
Improved this description and added couple of links related to dataflow templates